controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+                                                  {0.867739, -0.343277, -0.359429, 1.75753}};                               
+
+
+/* This is the main program of the supervisor. It creates an instance of the Supervisor
+ *     instance, launches its function(s) and destroys it at the end of the execution.
+ */
+int main() {
+    std::cout << "a" << std::endl;
+    // create the Supervisor instance and assign it to a robot
+    Supervisor* supervisor = new Supervisor();
+    Node* target           = supervisor->getFromDef("RED_1");
+    std::cout << "b" << std::endl;
+    // Get the time step of the current world.
+    int timeStep = (int) supervisor->getBasicTimeStep();
+
+    std::cout << "c" << std::endl;
+
+    // Add the motion manager to make the robot stand up and look around
+    managers::RobotisOp2MotionManager* mMotionManager = new managers::RobotisOp2MotionManager(supervisor);
+    mMotionManager->playPage(54);
+
+    // Generate random seed
+    srand(time(NULL));
+
+    // Main loop:
+    while (supervisor->step(timeStep) != -1) {
+
+        // Grab the current translation field of the robot to modify
+        Field* target_translation_field = target->getField("translation");
+        // Convert the field to a vector to output to console
+        const double* target_translation_vec = target_translation_field->getSFVec3f();
+
+        // Output current location
+        std::cout << "Location: " << std::endl;
+        std::cout << "X: " << target_translation_vec[0];
+        std::cout << " Y: " << target_translation_vec[1];
+        std::cout << " Z: " << target_translation_vec[2] << std::endl;
+
+        // Prepare new location
+
+        /* 0, 0, 0 is centre of the playing field.
+         *     X ranges from -5.4 - 5.4, the positive value on the left hand side when looking
+         *     from red
+         *     Y ranges from -8 - 8, the positive value on the red goal side
+         *     Z should be set to 3.2 to be on the ground
+         *     If the robot teleports into an existing object it may run into issues for that
+         *     image only, after resetPhysics is should return to a regular state
+         */
+
+        double newPos[3];
+        newPos[0] = 5.4 - (rand() % 1080) / 100;  // target_translation_vec[0] + 1;
+        newPos[1] = 8 - (rand() % 1600) / 100;    // target_translation_vec[1] + 0;
+        newPos[2] = 0.32;                         // target_translation_vec[2] + 0;
+
+        // Set new location
+        target_translation_field->setSFVec3f(newPos);
+
+        // Grab the current rotation field of the robot to modify
+        Field* target_rotation_field = target->getField("rotation");
+        // Convert the field to a vector to output to console
+        const double* target_rotation_vec = target_rotation_field->getSFRotation();
+
+        // Loop 4 times, one for each direction to face
+        for (int i = 0; i < rotationsSize; i++) {
```

Comment:
Using a `std::array` or `std::vector` allows us to use range-based for loops. These are basically always preferred to manually setting dummy variables for for loops, like here.

Reply:
Link is dead :( But yeah I am assuming that it would be similar to a python list having a size stored that I can access rather than having to save it separately

Reply:
Here's an example of a range based for loop in our codebase:
https://github.com/NUbots/NUbots/blob/master/module/behaviour/Controller/src/Controller.cpp#L62

It also uses the `auto` keyword, which is very handy to know about as well.

Reply:
but yea, you've basically nailed it. It's a lot like `len(list)` in python, although it's a little more like if each list had a `.len()` method instead.

Reply:
Yep okidoki